# Week-1 Assignment

**Welcome to your first assignment for the SimuTech Winter Project 2022! I hope you are excited to implement and test everything you have learned up until now. There is an interesting set of questions for you to refine your acquired skills as you delve into hands-on coding and deepen your understanding of numpy, pandas, and data visualization libraries.**

P.S. You might find use of functions which you have never used before.So feel free to google

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Section0 : Importing Libraries

*Let's begin by importing numpy, pandas and matplotlib.*

In [5]:
#your code here
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Section1 : Playing with Python and Numpy

### Q1. Matrix Multiplication

##### (i) Check if matrix multiplication is valid

In [72]:
import numpy as np

def isValid(A, B):
  
  
  a,b=np.shape(A)
  c,d=np.shape(B)
  if(b==c):
    return True
  elif(a==b==c==d):
    return True
  else:
    return False 
  
 


##### (ii) Using loops (without using numpy)

In [92]:
def matrix_multiply(A,B):
    a,b=np.shape(A)
    c,d=np.shape(B)
    matrix=np.zeros((a,d))
    i=0
    while i<a:
        j=0
        while j<b:
            k=0
            while k<d:
                t=0
                t=t+(A[i][j]*B[j][k]) 
            k=k+1
        j=j+1
    i=i+1               
    matrix[i][k]=t      
    return matrix    
    c

##### (iii) Using numpy

In [48]:
def matrix_multiply_2(A,B):
  import numpy as np
  C=np.dot(A, B)
  print(C)


##### (iv) Testing your code

Run the following cell to check if your functions are working properly.

*Expected output:*
[ [102 108 114]
 [246 261 276]
 [390 414 438]
 [534 567 600] ]

In [93]:
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    [10, 11, 12]
])

B = np.array([
    [13, 14, 15],
    [16, 17, 18],
    [19, 20, 21]
])

if isValid(A,B):
  print(f"Result using loops: {matrix_multiply(A,B)}")
  print(f"Result using numpy: {matrix_multiply_2(A,B)}")
else:
  print(f"Matrix multiplication is not valid")

KeyboardInterrupt: 

### Q2. Z-Score Normalisation

Z-score normalization refers to the process of normalizing every value in a dataset such that the mean of all of the values is 0 and the standard deviation is 1.

We use the following formula to perform a z-score normalization on every value in a dataset:

New value = (x – μ) / σ

where:

x: Original value

μ: Mean of data

σ: Standard deviation of data

##### (i) Without using numpy

In [94]:
def mean(x):
  sum(x)/len(x)


In [95]:
def standard_deviation(x):
    return (sum((xi - mean(x))**2 for xi in x)/len(x))**0.5

  

In [99]:
def zscore_normalisation(x):
  return ((xi - mean(x))/standard_deviation(x) for xi in x)

##### (ii) Using numpy

Numpy has in_built functions for calculating mean and standard deviation

In [97]:
def zscore_normalisation_2(x):
  return (x-np.mean(x))/np.std(x)

##### (iii) Testing your code

Run the following cell to check if your functions are working properly.

*Expected Output:* [-1.06753267 -0.99745394 -0.99745394 -0.81057732 -0.41346451 -0.06307086
  0.31068237  0.91803138  1.22170588  1.89913361]

In [100]:
x = [4, 7, 7, 15, 32, 47, 63, 89, 102, 131]
print(f"Result without using numpy: {zscore_normalisation(x)}")
print(f"Result using numpy: {zscore_normalisation_2(x)}")

Result without using numpy: <generator object zscore_normalisation.<locals>.<genexpr> at 0x00000252DEAF5EE0>
Result using numpy: [-1.06753267 -0.99745394 -0.99745394 -0.81057732 -0.41346451 -0.06307086
  0.31068237  0.91803138  1.22170588  1.89913361]


### Q3. Sigmoid fn and its derivative

The sigmoid function is a mathematical function that maps any input value to a value between 0 and 1.

It is defined mathematically as s(x) = 1/(1+e^(-x)).

##### (i) Write a fn to implement sigmoid fn

In [101]:
def sigmoidfn(x):
  return 1/(1+np.exp(-x))

##### (ii) Write a fn to implement derivative of sigmoid fn

In [102]:
def derivative(x):
  return (1-sigmoidfn(x))*sigmoidfn(x)

##### (iii) Test your code

Run the following cell to check if your functions are working properly.

*Expected output:*

x on applying sigmoid activation fn is: [ [0.99987661 0.88079708 0.99330715 0.5        0.5       ]
 [0.99908895 0.99330715 0.5        0.5        0.5       ] ]

x on applying derivative of sigmoid activation fn is: [ [-1.23379350e-04 -1.04993585e-01 -6.64805667e-03 -2.50000000e-01
  -2.50000000e-01]
 [-9.10221180e-04 -6.64805667e-03 -2.50000000e-01 -2.50000000e-01
  -2.50000000e-01] ]

In [103]:
x = np.array([
    [9,2,5,0,0],
    [7,5,0,0,0]
])
print(f"x on applying sigmoid activation fn is: {sigmoidfn(x)}")
print(f"x on applying derivative of sigmoid activation fn is: {derivative(x)}")

x on applying sigmoid activation fn is: [[0.99987661 0.88079708 0.99330715 0.5        0.5       ]
 [0.99908895 0.99330715 0.5        0.5        0.5       ]]
x on applying derivative of sigmoid activation fn is: [[1.23379350e-04 1.04993585e-01 6.64805667e-03 2.50000000e-01
  2.50000000e-01]
 [9.10221180e-04 6.64805667e-03 2.50000000e-01 2.50000000e-01
  2.50000000e-01]]


# Section 2: Playing with Pandas

###Read the csv file


In [108]:
import pandas as pd
df = pd.read_csv("superheated_vapor_properties.csv")
df

Pressure Property    Liq_Sat      Vap_Sat           75          100  \
0         1.0        V     1.0000  129200.0000  160640.0000  172180.0000   
1         1.0        U    29.3340    2385.2000    2480.8000    2516.4000   
2         1.0        H    29.3350    2514.4000    2641.5000    2688.6000   
3         1.0        S     0.1060       8.9767       9.3828       9.5136   
4        10.0        V     1.0100   14670.0000   16030.0000   17190.0000   
..        ...      ...        ...          ...          ...          ...   
539   11200.0        S     3.4440       5.5476          NaN          NaN   
540   11400.0        V     1.5040      15.2840          NaN          NaN   
541   11400.0        U  1450.0000    2527.2000          NaN          NaN   
542   11400.0        H  1467.2000    2701.5000          NaN          NaN   
543   11400.0        S     3.4575       5.5357          NaN          NaN   

             125          150          175          200  ...        425  \
0    183720.0000  195270.0000  206810.0000  218350.0000  ...        NaN   
1      2552.3000    2588.5000    2624.9000    2661.7000  ...        NaN   
2      2736.0000    2783.7000    2831.7000    2880.1000  ...        NaN   
3         9.6365       9.7527       9.8629       9.9679  ...        NaN   
4     18350.0000   19510.0000   20660.0000   21820.0000  ...        NaN   
..           ...          ...          ...          ...  ...        ...   
539          NaN          NaN          NaN          NaN  ...     6.2491   
540          NaN          NaN          NaN          NaN  ...    24.1040   
541          NaN          NaN          NaN          NaN  ...  2873.1000   
542          NaN          NaN          NaN          NaN  ...  3147.9000   
543          NaN          NaN          NaN          NaN  ...     6.2367   

             450        475          500        525          550        575  \
0    333730.0000        NaN  356810.0000        NaN  379880.0000        NaN   
1      3049.9000        NaN    3132.4000        NaN    3216.7000        NaN   
2      3383.6000        NaN    3489.2000        NaN    3596.5000        NaN   
3        10.8200        NaN      10.9612        NaN      11.0957        NaN   
4     33370.0000        NaN   35670.0000        NaN   37980.0000        NaN   
..           ...        ...          ...        ...          ...        ...   
539       6.3515     6.4452       6.5324     6.6147       6.6929     6.7679   
540      25.5990    27.0100      28.3590    29.6610      30.9250    32.1600   
541    2929.4000  2982.6000    3033.9000  3083.9000    3133.1000  3181.9000   
542    3221.2000  3290.5000    3357.2000  3422.1000    3485.7000  3548.5000   
543       6.3399     6.4341       6.5218     6.6043       6.6828     6.7580   

             600        625          650  
0    402960.0000        NaN  426040.0000  
1      3302.6000        NaN    3390.3000  
2      3705.6000        NaN    3816.4000  
3        11.2243        NaN      11.3476  
4     40290.0000        NaN   42600.0000  
..           ...        ...          ...  
539       6.8401     6.9099       6.9777  
540      33.3700    34.5600      35.7330  
541    3230.4000  3278.8000    3327.2000  
542    3610.8000  3672.8000    3734.6000  
543       6.8304     6.9004       6.9683  

[544 rows x 37 columns]

###Display the shape of data frame


In [109]:
data=pd.DataFrame(df)
data.shape


(544, 37)

###Return an array containing names of all the columns

In [110]:
np.array(data.columns)

array(['Pressure', 'Property', 'Liq_Sat', 'Vap_Sat', '75', '100', '125',
       '150', '175', '200', '220', '225', '240', '250', '260', '275',
       '280', '290', '300', '320', '325', '340', '350', '360', '375',
       '380', '400', '425', '450', '475', '500', '525', '550', '575',
       '600', '625', '650'], dtype=object)

###Display the number of null values in each column of the dataframe``



In [111]:
data.isnull()

Pressure  Property  Liq_Sat  Vap_Sat     75    100    125    150    175  \
0       False     False    False    False  False  False  False  False  False   
1       False     False    False    False  False  False  False  False  False   
2       False     False    False    False  False  False  False  False  False   
3       False     False    False    False  False  False  False  False  False   
4       False     False    False    False  False  False  False  False  False   
..        ...       ...      ...      ...    ...    ...    ...    ...    ...   
539     False     False    False    False   True   True   True   True   True   
540     False     False    False    False   True   True   True   True   True   
541     False     False    False    False   True   True   True   True   True   
542     False     False    False    False   True   True   True   True   True   
543     False     False    False    False   True   True   True   True   True   

       200  ...    425    450    475    500    525    550    575    600  \
0    False  ...   True  False   True  False   True  False   True  False   
1    False  ...   True  False   True  False   True  False   True  False   
2    False  ...   True  False   True  False   True  False   True  False   
3    False  ...   True  False   True  False   True  False   True  False   
4    False  ...   True  False   True  False   True  False   True  False   
..     ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
539   True  ...  False  False  False  False  False  False  False  False   
540   True  ...  False  False  False  False  False  False  False  False   
541   True  ...  False  False  False  False  False  False  False  False   
542   True  ...  False  False  False  False  False  False  False  False   
543   True  ...  False  False  False  False  False  False  False  False   

       625    650  
0     True  False  
1     True  False  
2     True  False  
3     True  False  
4     True  False  
..     ...    ...  
539  False  False  
540  False  False  
541  False  False  
542  False  False  
543  False  False  

[544 rows x 37 columns]

###Create a column which contains the Pressure and Property columns, seperated with 'at' (For eg. V at 1, H at 101.325). Using this print the following:
- Enthalpy at 75 kPa and 573 K
- Entropy at 493 K and 250 kPa



In [125]:
df['Property at Pressure']=df['Property'] + " at " + df['Pressure'].astype(str)

idx1=df[df['Property at Pressure']=='H at 75.0'].index
k=df['300'].loc[idx1].item()
print(f"{k}")

idx2=df[df['Property at Pressure']=='S at 250.0'].index
l=df['220'].loc[idx2].item()
print(f"{l}")

3075.1
nan


###Find out the column with the highest number of missing values

NameError: name 'nan' is not defined

###What is the average enthalpy of Sat. Liq. at all different pressures in the dataset ?

In [132]:
print(df['Liq_Sat'].iloc[2::4].mean())

936.9707720588235


###Separate out the V,U,H,S data from the dataset into V_data, U_data, H_data, S_data

In [133]:
grouped = data.groupby(data.Property)
V_data = grouped.get_group("V")
U_data = grouped.get_group("U")
H_data = grouped.get_group("H")
S_data = grouped.get_group("S")

#Section 3: PLotting the PLOTS

###Plot the properties (specific volume, specific internal energy, specific enthalpy, specific entropy) vs Pressure for saturated liquid.

Note:
- Try using the subplot feature of matplotlib(Explore it!!)
- Provide appropriate title, labels, markersize and other parameters to the plot

In [ ]:
#Need some more time to understand matplotlib.

###Plot the specific volume of saturated liquid between 300 kPa and 1500 kPa

In [ ]:
#your code here

#Linear Regression
You're given a 100 data points generated from the function
$y=mx+c$. Some noise is added to the y values.

**Task:** Estimate the values of m and c using linear regression. Write a suitable cost function and update the parameters.

In [135]:
 # Generation of data points
x = np.random.rand(100)*100
y = -4*x+15
y = y+np.random.randn(100)*30
plt.scatter(x,y)

NameError: name 'plt' is not defined

In [ ]:
# Initialisng parameters m and c such that the best fit line is given by y = m*x+c

#Cost function to be minimised

#Function for updating parameter


Run the model

In [ ]:
# define number of iterations for minimising cost function

# define the learning rate #

# run the model

In [ ]:
print(m,c)  # value should be close to (-4,15)